contracts/MarketOracle.sol

```diff
@@ -14,104 +14,128 @@ interface IOracle {
 /**
  * @title Market Oracle
  *
- * @dev Provides the aggregated exchange rate data onchain using data from a whitelisted
- *      set of market source contracts.
- *      Exchange rate is the TOKEN:TARGET rate.
+ * @dev Provides the aggregated rate data onchain using data from a whitelisted
+ *      set of providers.
  */
 contract MarketOracle is Ownable, IOracle {
     using SafeMath for uint256;
 
-    // Whitelist of sources
-    MarketSource[] public whitelist;
+    struct Report {
+        address provider;
+        uint256 timestamp;
+        uint256 payload;
+    }
 
-    event LogSourceAdded(MarketSource source);
-    event LogSourceRemoved(MarketSource source);
-    event LogSourceExpired(MarketSource source);
+    mapping (address => uint256) internal whitelistMap;
+    Report[] public whitelistReports;
+
+    event LogSourceAdded(address source);
+    event LogSourceRemoved(address source);
+    event LogSourceExpired(address source);
+
+
+    // The number of seconds after which the report must be deemed expired.
+    uint256 public reportExpirationTimeSec = 6 hours;
+
+    uint256 public minimumProviders = 1;
+
+    function setReportExpirationTimeSec(uint256 reportExpirationTimeSec_)
+    external
+    onlyOwner
+    {
+        require(reportExpirationTimeSec_ > 0);
+        reportExpirationTimeSec = reportExpirationTimeSec_;
+    }
+
+    function setMinimumProviders(uint256 minimumProviders_)
+    external
+    onlyOwner
+    {
+        require(minimumProviders_ > 0);
+        minimumProviders = minimumProviders_;
+    }
+
+    function lookupProviderIndex(address provider) internal view returns (uint256) {
+        uint256 position = whitelistMap[provider];
+        require(position != 0);
```

Comment:
Would prefer moving this `require` to a function modifier (`onlyWhitelistedProvider`) which can be used for authorization and `lookupProviderIndex` to not fail at all.

Reply:
Unless the same logic is used more than one place, I'd prefer to keep it inline too.